### ISSUE TITLE
Handle partial responses when the server returns an error

### ISSUE THREAD

**Issue by stubailo** [apollographql/apollo-client#51]
`XXX` in `QueryManager#makeRequest`.

**Comment by stubailo** [comment]
Still haven't fixed this

**Comment by stubailo** [comment]
Interesting news: Relay doesn't do anything about this. They just ignore the error and put the incorrect data into the store. Perhaps this is because the Facebook backend doesn't return errors like that?

**Comment by martijnwalraven** [comment]
Hmmm, so if the server sends a partial response plus errors, you just receive the partial response?

Do you know how they deal with https://github.com/apollostack/apollo-client/issues/106?

**Comment by stubailo** [comment]
They just overwrite the store with 'null' so you can't actually tell if something is missing due to an error or if it's actually missing. I asked them about this specifically.

**Comment by stubailo** [comment]
Marking this as low priority because Relay doesn't do this, and GraphQL-JS doesn't even have a way to get this information.

**Comment by BartKrol** [comment]
I've started playing with `apollo-client` and run into a problem which I guess it's already discussed here (not sure though - I didn't want to open another issue :P)

I'm making a following query to my graphql server with GraphqiQL: 
```
{
  team {
    name,
    members {
      name
    }
  }
}
```
As a result I'm getting some data and an Error, because one team member had an incorrect id. It's not ideal, but I consider this as a valid response. Unfortunately when I make the same query with `apollo-client` I'm getting only a list of `ApolloErrors`. In my opinion this is an incorrect behaviour.

**Comment by stubailo** [comment]
> In my opinion this is an incorrect behaviour.

Relay will do something similar - basically, since errors don't come with a reliable location there is no way to know which part of the response is reliable. This is something we're actively trying to get fixed in GraphQL itself.

But perhaps we could add an option to be a bit looser with error handling.

**Comment by nmuth** [comment]
+1 for a config option. I would like to be able to deal with incorrect data myself while the proper GraphQL solution is being developed.

**Comment by stubailo** [comment]
So here's where that gets handled: https://github.com/apollostack/apollo-client/blob/10cfa7962e58f226573ae6eb63f30c2ce528d7ea/src/QueryManager.ts#L273

What should we do instead?

**Comment by nmuth** [comment]
@stubailo Thanks for the context-- I'll take a look tonight and give you my thoughts tomorrow.

**Comment by zhewison** [comment]
👍  For this to.
Being able to use the errors to decide what to do with the data that _is_ returned will be helpful.

**Comment by nmuth** [comment]
At a high level, what would be ideal for our use case would be an option passed to `ApolloClient`:

```
const client = new ApolloClient({
  partialResultsOnError: true
})
```

When this option is set, and a query contains errors, the client would resolve successfully with what information it _did_ receive and attach an `errors` field:

```
{
  data: [{
    team: {
      members: [{
        id: 'b',
        name: 'Foo Bar'
      }, null, null]
  }],
  errors: [ Error(could not find id 'a'), Error(could not find other id 'c') ]
}
```

If `partialResultsOnError` is `false`, the client would act exactly as it does now.

How well does this fit into the existing system?

**Comment by stubailo** [comment]
So here's the long and short of the problem.

1. Apollo uses a global cache to store all of the data you have ever fetched, which is important if you want your UI to be consistent, or you need to handle mutation results and have them show up properly.
2. GraphQL.js simply returns `null` on any field where there is an error, and doesn't have any indication of where in the response the error happened.

So, basically this unfortunate situation can happen:

1. You fetch query A, the result is cached
2. You fetch query A again, but this time there's an error that caused a `null`, the result is cached
3. Now, you have a `null` in your cache, even thought you had some good data from (1)

Given that there's no way to tell the difference between a real null result and an error, we decided it's best to just keep the original data and not let the new result "poison" it.

What's a different approach that would be better?

For example, we could just always put the data in the cache and deal with the fact that it might be wrong; we could have a configuration option that would let you look at the error and say whether the result should be used or not; maybe something else?

**Comment by BartKrol** [comment]
I don't know how caching is done in `apollo`, but in the example I've shown I'd like to render as much as I can for my UI. When making a request for a `team` and `team.members`, I'd like to show all the information about a team and use data about those people I could get from `team.members`. With current state when one members fails I cannot use any of the data.

With this being said I think it's a good approach to put all of the data into the cache even if it's invalid or at least make an option to mark queries which might fail (and put them always in a cache).

**SideNote** Regarding the second point:

> GraphQL.js simply returns null on any field where there is an error, and doesn't have any indication of where in the response the error happened.

For this example query
```
{
  team {
    name,
    members {
      name
    }
  }
}
```
I'm getting:
```
...
    {
      "message": "Error: cannot GET http://peopleapi.com/people/testID (404)",
      "locations": [
        {
          "line": 5,
          "column": 6
        }
      ]
    },
```

Which, as far as I understand, will show where error happened in the query. Of course it doesn't tell everything, because for example it will not tell which `member` exactly it was, but it might be interesting to use it to invalidate only a part of a query by apollo.

**Comment by stubailo** [comment]
We could take advantage of this PR we added to graphql-js, to know _exactly_ where the error happened: https://github.com/graphql/graphql-js/pull/396

But yeah, I'd be OK to add an option to ignore the errors and just put everything into the cache until reporting the error path is part of the GraphQL spec.

Do we want it to be per-query, or global?

**Comment by nmuth** [comment]
Per-query is probably more user-friendly as it gives the finest level of control. But maybe there could be a global option as well which would provide a default. I'm suggesting this because personally I would like to turn on a global option and not worry about it, but other people might need to cache more aggressively or maintain functionality without a network connection.

Is there currently a pattern for per-query options with a global default?

**Comment by stubailo** [comment]
No - let's add the thing that you need, and if someone comes along and needs per-query control then we'll add it later. I'd rather not add hypothetical features.

**Comment by nmuth** [comment]
Sounds good. One more question: should this behavior apply to mutations as well? We're not using mutations yet so I'm less familiar with that area.

**Comment by stubailo** [comment]
Yes, mutation results are handled pretty much the same way. One thing that's unfortunate is, it would be good to differentiate an error that caused the data to not be saved, from an error in the subsequent refetch. But I don't think GraphQL lets us do that today.

**Comment by BartKrol** [comment]
@stubailo do you think it's better to leave a fix in your hands or should one of us create a PR (@nmuth, @zhewison or myself)? It's not super urgent, but it would be definitely nice to have it.

**Comment by helfer** [comment]
@BartKrol @nmuth @zhewison We love PRs on all our repos, so one of you should definitely create a PR! :tada:

**Comment by helfer** [comment]
Initially we'll just provide an option to choose between writing or not writing the result into the store when an error happens.

**Comment by zhewison** [comment]
@stubailo @helfer 
Awesome that you've included this in the next milestone!

We're locked into `v0.4.20` for now and since `master` is very much diverged from that; I've created a patch for `v0.4.20`.

The `returnErrorsWithResults` configuration is currently not covered by the tests in this branch as i'm conflicted on the longevity of `v0.4.20`.

https://github.com/apollostack/apollo-client/compare/apollostack:v0.4.20...zhewison:errors-with-response

Can you let me know if this is helpful or should be implemented in `master`? I'm unsure on the timeline for the next milestone.

I can open a PR if you have any comments. Thanks!

**Comment by andela-iamadi** [comment]
Just hit a glitch with this issue too. Apollo reject all GraphQL response because of one error. Where are we with this?

Anyone jumping on it yet?

**Comment by helfer** [comment]
Error handling is high on our list of priorities, but we decided not to do it for 0.5 yet.

**Comment by marktani** [comment]
Any news on this? This is pretty important in my opinion :slightly_smiling_face:

**Comment by helfer** [comment]
@marktani it's not high on our list of priorities, but we're definitely open to a PR. What makes this tricky is having predictable and identical results when hitting the cache vs. requesting from the server. I don't really know what the right solution there is, so it would help to have a couple of real-world use-cases so we can weigh the pros and cons of each option. Maybe you could describe your use-case here for a start! 🙂

**Comment by bl42** [comment]
`
{
  member { # resolves from DB
    firstName
    ...
    tweet(latest:true) { # resolves from twitter api
       ... 
    }
  }
}
`
If an external api fails, it doesnt fail gracefully where i think we all in this use case would want it too

**Comment by ali** [comment]
I came up with a workaround by creating a custom network interface: https://gist.github.com/ali/4884425affbe43f9db85726fecf7d033

Since a partial response is a response with both data and errors, my solution was to wrap Apollo's network interface and make it ignore errors. 😛 I know, unspeakable. It'd be nice if I could additionally retry the request (or just the parts of the query that failed) so that I could potentially update the store with the full response behind the scenes.

Here's our use case: Our GraphQL service calls out to another service to resolve the `isOnline` field on our `User` type. Sometimes this part of the query will fail, and I'll get a response with all our users' data except for that `isOnline` field. I'd like to be able to still render these users and display them as offline until I can refetch their online status.

My actual workaround in production is to use a separate container and query just for that `isOnline` and nest it inside containers that query for the rest of a user's fields.

**Comment by marktani** [comment]
> Maybe you could describe your use-case here for a start!

My use case is any kind of error that happen after/during the query resolving. So it's a valid query as far as the GraphQL type system can tell, but another layer determined it's an error.

In the case of Graphcool, this is mostly permission errors - a node or field was included in the query that the requestee has no permission to access. [Here's an example query](https://www.graph.cool/docs/faq/graphcool-insufficient-permissions-query-response-noohie9hai/#an-optional-field-in-the-query-is-not-accessible):

```graphql
query {
  allDocuments {
    public
    optionalSecret
  }
}
```

```
{
  "data": {
    "allDocuments": [
      {
        "public": "Everyone can read this!",
        "optionalSecret": null
      },
      {
        "public": "And this as well",
        "optionalSecret": null
      }
    ]
  },
  "errors": [
    {
      "locations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "path": [
        "allDocuments",
        0,
        "optionalSecret"
      ],
      "code": 3008,
      "message": "Insufficient Permissions",
      "requestId": "cizfflsqy5n4q0188qhz66euh"
    },
    {
      "locations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "path": [
        "allDocuments",
        1,
        "optionalSecret"
      ],
      "code": 3008,
      "message": "Insufficient Permissions",
      "requestId": "cizfflsqy5n4q0188qhz66euh"
    }
  ]
}
```

**Comment by cfatorNSS** [comment]
+1 on this.  One easy option might be to support a .finally(response) block on the query promises.

**Comment by reut-co** [comment]
hey, any news regarding this issue?

**Comment by claudiamatosa** [comment]
This would be quite important for a project I'm working on too. Use case (dummy schema):

```
getCustomers {
  name
  email
  favouriteBook {
    id
    name
  }
}
```

`favouriteBook` needs to be fetched with a resolver for each `Customer` in the list, via http, as it's in a separate micro-service. If one of the resolvers fails, meaning that we don't get a group for that customer, apollo client doesn't return the remaining data, even though it comes in the network response. This means that the entire list fails to display, even though only one of the elements had an error.

Now, maybe we can handle this in the resolver by catching the error and returning empty data, however it would still be helpful to see the errors, as taking this approach would completely hide potential problems.

**Comment by tmickel** [comment]
@helfer @stubailo I'm still learning about this problem (and Apollo), but:

In an imaginary Apollo configuration where partial results and errors are returned, could these be cached, too? I.e., a cached response would indicate errors (and partial data), which would then be overwritten (errors cleared) if a new, successful response comes back from the server?

In react-apollo, errors could maybe even be passed to a component as a prop in this mode?

**Comment by sorenbs** [comment]
I have been living in blissful ignorance for the last 6 months thinking this had already been fixed :-)

The GraphQL spec is pretty explicit about supporting partial responses and the errors field containing detailed information.

As the most popular GraphQL client I would expect Apollo to have a strong story around error handling.

@helfer @stubailo - what can Graphcool do to help move this forward?

**Comment by helfer** [comment]
@sorenbs this is currently blocked on the store refactor which we're working on at the moment. Currently the plan for the new API is to handle errors the same way as we handle data, and write them into the store as well if they have a path (which means you might retrieve cached errors from the store). I'm not yet 100% clear on what we should do with errors that don't have a path, but my intuition is to give people a lot of control around how they want to handle those, and decide on a per-query basis whether something with errors should be written to the store or returned from the store. This will make it very easy to configure Apollo Client in just the way people need, but it comes with the drawback that it will be more complicated to tell what's really going on, which is why choosing sensible defaults will be very important.

What you could help us with is letting us know exactly how you would want your client to behave in different circumstances, and helping us choose sensible defaults so that Apollo Client does what people expect right out of the box (while keeping options open to modify its behavior).

Here are some questions to discuss:
1. Should errors (with/without path) be written to the store?
2. Should errors be retrieved from the cache if an earlier query errored on the same path?
3. What should happen with queries that have errors without path?
4. How should GraphQL errors be delivered to components? How about non-GraphQL errors?
5. If we don't cache errors, how should we deal with them during SSR?

There are probably a few more missing here, but

**Comment by sorenbs** [comment]
Thanks Jonas - that's awesome news!

I very much agree with allowing developers to configure the behaviour, and agree the sensible defaults are super important.

I'll take some time soon to write up my perspective on this.

**Comment by thifranc** [comment]
Hi !  

@helfer  thanks a lot for all the work you do, that's amazing !
Do you have any leads about when the release of the "error handler feature" for apollo-client is due ?
And moreover, we're currently implementing an error handling system following [this article](https://medium.com/@tarkus/validation-and-user-errors-in-graphql-mutations-39ca79cd00bf). 
 
If we return something like :  
```
{
  errors: [{
    message: 'Request is invalid.',
    path: ['createUser'],
    myCustomErrorData: {
      '': ['Failed to create a new user.'],
      email: ['A user with this email already exists.']
    },
    ...
  }]
}
```
Do you think that 'myCustomErrorData' will be accessible with the apollo-client's latest release ?

Cheers,

**Comment by cesarsolorzano** [comment]
@thifranc you could use `response.graphQLErrors[0].myCustomErrorData` to access the data.

**Comment by oliverjam** [comment]
Can we re-open this? The problem still exists—I just hit it today.

Querying event data means that for some languages e.g. a town might not exist and returns null.

```graphql
query eventQuery {
  transaction(id: 111111) {
    performance(siteId: 1) {
      venue {
        name
        town
      }
    }
  }
}
```
We get this response:
```json
{
  "data": {
    "transaction": {
      "performance": {
        "venue": {
          "name": "Kobetamendi",
          "town": null,
        }
      }
    }
  },
  "errors": [
    {
      "message": "Error trying to resolve town.",
      "locations": [
        {
          "line": 9,
          "column": 11
        }
      ]
    }
  ]
}
```
Apollo doesn't pass any of the partial data it gets right now, which is frustrating.

**Comment by larryranches** [comment]
@helfer Is there any traction on this?

**Comment by vidyakhadsareibm** [comment]
@oliverjamI am facing the exact same issue you mentioned. @stubailo  do you have any updates?

**Comment by stubailo** [comment]
@jbaxleyiii has this been resolved with the new error policies? BTW we might want to add a message to the bot that says you can make it not stale by commenting

**Comment by vidyakhadsareibm** [comment]
@jbaxleyiii please let us know if this is fixed. Its affecting my design decisions.

**Comment by bl42** [comment]
https://www.apollographql.com/docs/react/basics/queries.html#graphql-config-options-errorPolicy

This has solved all my problems

**Comment by tmickel** [comment]
awesome!!!!!! thank you!!

**Comment by liviudobrea** [comment]
I'm getting the same _unexpected_ behaviour as @AndrewHenderson with using **`errorPolicy: all`**.
The network request contains both `data` and `errors` properties, but somehow `react-apollo` skips `errors` when merging the result passed to the component props.
However, I am able to catch said errors via middleware.
One (innapropiate) alternative I found working for now is manually requesting the query and processing the callbacks:
```
this.props.client.query({
        query: getVendor,
	variables: { id },
	errorPolicy: 'all'
}).then(({ data, loading, errors }) => {
	// Props are available here, but are not assigned to the component props
        // when using the graphql decorator
})
```

**Comment by hwillson** [comment]
Hi all - a lot has changed since this issue was first created. Apollo Client now has much better [error handling capabilities](https://www.apollographql.com/docs/react/features/error-handling.html). I'll close this issue for now since most of the comments here are outdated, but please open up new issues if you run into further error handling problems with current day versions of `apollo-client`. Thanks!
